velox/dwio/dwrf/writer/WriterContext.h

```diff
@@ -311,10 +313,17 @@ class WriterContext : public CompressionBufferPool {
   }
 
   int64_t getEstimatedOutputStreamSize() const {
+    // NOTE: This is a rough heuristic, based on buffer growth factors. Due to
+    // the chained/paged buffer semantics of output streams, the larger the
+    // streams are, the more we are undercounting content size. However, this
+    // ensures that we don't produce small stripes for flat map scenarios. A
+    // better heuristics would have to account for the number of streams in some
+    // way.
     return (int64_t)std::ceil(
-        (getMemoryUsage(MemoryUsageCategory::OUTPUT_STREAM) +
-         getMemoryUsage(MemoryUsageCategory::DICTIONARY)) /
-        getConfig(Config::COMPRESSION_BLOCK_SIZE_EXTEND_RATIO));
+        (getMemoryUsage(MemoryUsageCategory::OUTPUT_STREAM) /
+             getConfig(Config::COMPRESSION_BLOCK_SIZE_EXTEND_RATIO) +
+         getMemoryUsage(MemoryUsageCategory::DICTIONARY) /
+             kDataBufferGrowthFactor));
```

Comment:
We don't compress dictionary memory? and kDataBufferGrowthFactor only applies for dictionary buffer allocation? Thanks!

Reply:
See my comment above this method. 

The underlying logic is to estimate the content size, given the memory utilization. Output stream uses a chained buffer semantic with a configurable growth ratio for the head buffer, whereas dictionary values/alphabets uses a hardcoded growth factor in DataBuffer class, hence this equation. Both terms are calculated conservatively to ensure we write big enough stripes.